### PR TITLE
Prototyping new CodeAction API

### PR DESCRIPTION
**Pull request by mjbvz**
Adds skeleton of a new `CodeAction` type and allows codeActionProviders to return either `Command`s or `CodeAction`s

### CODE REVIEW

**Comment by mjbvz** [comment]
@jrieken Kai and I were talking today and wanted to try out a different API design than in #34664. The current change introduces two separate interfaces `QuickFix` and `Refactoring` instead of a single interface where quick fixes and refactorings are identified by a `type` field.

The implementation still need lots of work but please take a look at `vscode.proposed.d.ts` to see what you think of this proposal

**Comment by jrieken** [comment]
I like the general direction but not the change 🙊. I think the `Refactoring`-type is just a complicated way of expressing the same `Command` does, a bit of UI (label, tooltip), the actual command id, and some arguments. For the `Quickfix` I would use some composition, maybe

```ts
class QuickFix {
 command: Command
 addresses: Diagnostics[]
}
```

alternatively, in case we make add support for text edits, add a title

```ts
class QuickFix {
  title: string;
  fix: Command | WorkspaceEdit | TextEdit
  addresses: Diagnostics
}
```

or making it even more compact, with only one type 

```ts
class CodeAction {
 // ... all from above

 // when set treated as quick fix, otherwise as refactoring
 addresses?: Diagnostic[]
}
```

**Comment by jrieken** [comment]
Some inspiration from roslyn: http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/CodeFix.cs,7161867d561915e6. There seems to be a `QuickFix` type which is more or less a code action *with* a reference to the diagnostic it fixes.

**Comment by DustinCampbell** [comment]
In Roslyn...

* `CodeAction` is a named edit that can be invoked by the user. It contains a series of `CodeActionOperations` that can represent a text edit, a workspace edit, or something else. Often that "something else" is a dialog that allows the user to configure the `CodeAction` before invoking it. A great example of a `CodeAction` that requires a dialog is something like `Change Signature`.
* A `CodeFix` encapsulates a set of `CodeAction` objects with a `Diagnostic`.
* A `CodeRefactoring` encapsulates a set of `CodeAction` objects with a span in the editor, which could be collapsed as a caret position.

Does that help?

**Comment by mjbvz** [comment]
Thanks @DustinCampbell! Very helpful design to review. Do you happen to know how VS sets up keybindings for specific refactoring, such as `ctrl+r, m` for extract method? Is this done per-language or is there a general API for hooking these up?

**Comment by mjbvz** [comment]
@jrieken Today @kieferrm and I talked through the design again and cut it back to focus on address three problems:

1. No way to associate a quick fix with the diagnostic it addresses
1. No way to preview a quick fix's change
1. Refactoring actions are always calculated and shown with the lightbulb when just moving around in a file

The first two are addressed by adding a new `CodeAction` class.

For the third, one thought is to keep refactoring as code actions for now, but give code action providers additional context so that they can decide which actions to return. This latest change enables this by adding a `quickFixesOnly` flag on the `CodeActionContext`. When this flag is set, the refactoring provider would not return any code actions. The flag would always be set when just moving the cursor through code but unset if you manually request code actions with `cmd+.`.

**Comment by DustinCampbell** [comment]
> Do you happen to know how VS sets up keybindings for specific refactoring, such as ctrl+r, m for extract method? Is this done per-language or is there a general API for hooking these up?

There's no general API for this. It's done on a case-by-case basis.

File: src/vs/editor/contrib/quickFix/browser/quickFixCommands.ts [hunk 1]
```diff
@@ -39,13 +46,15 @@ export class QuickFixController implements IEditorContribution {
 	constructor(editor: ICodeEditor,
 		@IMarkerService markerService: IMarkerService,
 		@IContextKeyService contextKeyService: IContextKeyService,
-		@ICommandService commandService: ICommandService,
+		@ICommandService private readonly commandService: ICommandService,
```
Comment (jrieken) [inline src/vs/editor/contrib/quickFix/browser/quickFixCommands.ts]: `_` for privates please

File: src/vs/editor/common/modes.ts [hunk 2]
```diff
@@ -284,7 +292,7 @@ export interface CodeActionProvider {
 	/**
 	 * Provide commands for the given document and range.
 	 */
-	provideCodeActions(model: editorCommon.IReadOnlyModel, range: Range, token: CancellationToken): Command[] | Thenable<Command[]>;
+	provideCodeActions(model: editorCommon.IReadOnlyModel, range: Range, token: CancellationToken): (Command | CodeAction)[] | Thenable<(Command | CodeAction)[]>;
```
Comment (jrieken) [inline src/vs/editor/common/modes.ts]: In this land we can break the API so I'd suggest we don't carry along the `Command | CodeAction` notion but just make it a `CodeAction`. I'd also do that on the protocol layer and do the `Command` becomes `CodeAction` thing  in the code action adapter.

File: src/vs/editor/contrib/quickFix/browser/quickFix.ts [hunk 3]
```diff
@@ -7,16 +7,16 @@
...
 import { TPromise } from 'vs/base/common/winjs.base';
 import { onUnexpectedExternalError, illegalArgument } from 'vs/base/common/errors';
 import { IModelService } from 'vs/editor/common/services/modelService';
 import { CommonEditorRegistry } from 'vs/editor/common/editorCommonExtensions';
 
-export function getCodeActions(model: IReadOnlyModel, range: Range): TPromise<Command[]> {
+export function getCodeActions(model: IReadOnlyModel, range: Range): TPromise<(Command | CodeAction)[]> {
 
-	const allResults: Command[] = [];
+	const allResults: (Command | CodeAction)[] = [];
```
Comment (jrieken) [inline src/vs/editor/contrib/quickFix/browser/quickFix.ts]: `CodeAction` ftw

File: src/vs/editor/contrib/quickFix/browser/quickFixModel.ts [hunk 4]
```diff
@@ -112,11 +112,22 @@ export class QuickFixOracle {
 			const model = this._editor.getModel();
 			const range = model.validateRange(rangeOrSelection);
 			const position = rangeOrSelection instanceof Selection ? rangeOrSelection.getPosition() : rangeOrSelection.getStartPosition();
+
+			const fixes = getCodeActions(model, range).then(actions =>
+				actions.map(action => {
+					if ('id' in action) {
+						// must be a command
+						const command = action as Command;
+						return { title: command.title, command: command } as CodeAction;
```
Comment (jrieken) [inline src/vs/editor/contrib/quickFix/browser/quickFixModel.ts]: `CodeAction` ftw, lets do the conversion in the ext host

File: src/vs/editor/contrib/quickFix/browser/quickFixWidget.ts [hunk 5]
```diff
@@ -10,43 +10,38 @@ import { always } from 'vs/base/common/async';
...
 
-	constructor(editor: ICodeEditor, contextMenuService: IContextMenuService, commandService: ICommandService) {
-		this._editor = editor;
-		this._contextMenuService = contextMenuService;
-		this._commandService = commandService;
-	}
+	constructor(
+		private readonly editor: ICodeEditor,
+		private readonly contextMenuService: IContextMenuService,
+		private readonly onApplyCodeAction: (action: CodeAction) => TPromise<any>
```
Comment (jrieken) [inline src/vs/editor/contrib/quickFix/browser/quickFixWidget.ts]: unnecessary change, `_` for privates

File: src/vs/workbench/api/node/extHostLanguageFeatures.ts [hunk 6]
```diff
@@ -285,11 +285,37 @@ class QuickFixAdapter {
...
+
+				if (typeof action.command === 'string') {
+					return this._commands.toInternal(action as vscode.Command);
+				}
+
+				const codeAction = action as vscode.CodeAction;
+				return {
+					title: codeAction.title,
+					command: codeAction.command ? this._commands.toInternal(codeAction.command) : undefined,
+					edits: codeAction.edits
```
Comment (jrieken) [inline src/vs/workbench/api/node/extHostLanguageFeatures.ts]: For the sake of simplicity we can also make `edits` always a `WorkspaceEdit`. We'd treat the `TextEdit` as sugar and translate it into a workspace edit... Makes the internal CodeAction simpler and easier to handle.

**Review by jrieken — CHANGES_REQUESTED** [review]

**Comment by jrieken** [comment]
I would have gone ahead but there are some merge-conflicts...

File: src/vs/editor/common/modes.ts (hunk 2 above)
Comment (mjbvz) [reply src/vs/editor/common/modes.ts]: I'm concerned this will break the `vscode.executeCodeActionProvider` command. I believe this should still return commands instead of code actions for old providers and so  I'm going to revert the change to only use code actions for now
